funsor/terms.py

```diff
@@ -20,6 +20,63 @@
 from funsor.six import getargspec, singledispatch
 
 
+@dispatch(object, object)
+def substitute(expr, subs):
+    return expr  # default: do nothing
+
+
+@substitute.register(object, tuple)
+def substitute_with_tuple(expr, subs):
+    return substitute(expr, dict(subs))
+
+
+@substitute.register(tuple, dict)
+def substitute_tuple(expr, subs):
+    return tuple(substitute(v, subs) for v in expr)
+
+
+@substitute.register(Domain, dict)
+def substitute_domain(expr, subs):
+    return expr
+
+
+@substitute.register(frozenset, dict)
+def substitute_frozenset(expr, subs):
+    return frozenset(substitute(v, subs) for v in expr)
+
+
+@substitute.register(OrderedDict, dict)
+def substitute_ordereddict(expr, subs):
+    return OrderedDict([(k, substitute(v, subs)) for k, v in expr.items()])
+
+
+def alpha_convert(expr):
+    if not expr.bound or all("__BOUND" in name for name in expr.bound):
+        return expr
+
+    alpha_subs = {name: interpreter.gensym(name + "__BOUND")
+                  for name in expr.bound if "__BOUND" not in name}
+
+    new_values = []
+    for v in expr._ast_values:
+        v = substitute(v, alpha_subs)
+        if isinstance(v, str):
+            v = alpha_subs[v] if v in alpha_subs else v
+        elif isinstance(v, frozenset):
+            swapped = v & frozenset(alpha_subs.keys())
+            v |= frozenset(alpha_subs[k] for k in swapped)
+            v -= swapped
+        elif isinstance(v, tuple) and isinstance(v[0], tuple) and len(v[0]) == 2 and \
+                isinstance(v[0][0], str) and isinstance(v[0][1], Funsor):
+            v = tuple((alpha_subs[k] if k in alpha_subs else k, vv) for k, vv in v)
+        elif isinstance(v, OrderedDict):  # XXX is this case ever actually triggered?
+            v = OrderedDict([(alpha_subs[k] if k in alpha_subs else k, vv) for k, vv in v.items()])
+        new_values.append(v)
+
+    # TODO should this call reflect explicitly?
```

Comment:
Yes, I think this should call `reflect()` explicitly.

It already feels dangerous to make `reflect()` reentrant, and making it subject to interpretation seems even more dangerous. If we assume interpretations are alpha-invariant, then it should be fine to call `reflect()` here. The only issue I can possibly think of is when running under some sort of tracing interpretation.